cpp/src/arrow/engine/substrait/expression_internal.cc

```diff
@@ -320,6 +320,30 @@ Result<compute::Expression> FromProto(const substrait::Expression& expr,
       return function_converter(substrait_call);
     }
 
+    case substrait::Expression::kCast: {
+      const auto& cast_exp = expr.cast();
+      ARROW_ASSIGN_OR_RAISE(auto input,
+                            FromProto(cast_exp.input(), ext_set, conversion_options));
+
+      // need to convert from the Substrait type spec to the relevant Arrow thing
+      // do we need a switch/case or can we just get code similar to the below FromProto?
+      // the type should come from cast_expr.type()
+      
+      // this will be a Result<std::pair<std::shared_ptr<DataType>
+      ARROW_ASSIGN_OR_RAISE(auto type,
+                            FromProto(cast_exp.type(), ext_set, conversion_options));
+
+      if(cast_exp.failure_behavior() == substrait::Expression_Cast_FailureBehavior::Expression_Cast_FailureBehavior_FAILURE_BEHAVIOR_THROW_EXCEPTION){
+        return compute::call("cast", std::move(input), compute::CastOptions::Safe(type));
+      } else if(cast_exp.failure_behavior() == substrait::Expression_Cast_FailureBehavior::Expression_Cast_FailureBehavior_FAILURE_BEHAVIOR_RETURN_NULL){
+
+      // e.g. if unspecified  
+      } else {
+        return compute::call("cast", std::move(input), compute::CastOptions::Unsafe(type));
```

Comment:
I notice you switch between safe and unsafe here.  I'm not sure safe/unsafe (i.e. is this value invalid?) is the same as throw/null (i.e. this value is invalid, what should we do?).  For example, an unsafe cast from int16 to int8 would overflow 500 to -12 instead of assigning null.  It seems that Substrait has no concept of safe/unsafe (might be a good addition to the spec).

Reply:
Thanks for highlighting this, I can see that these are different things, so I'll update this.